lalrpop/src/grammar/parse_tree.rs

```diff
@@ -375,11 +377,22 @@ pub struct MacroSymbol {
 
 impl TerminalString {
     pub fn quoted(i: InternedString) -> TerminalString {
-        TerminalString::Literal(TerminalLiteral::Quoted(i))
+        TerminalString::Literal(TerminalLiteral::Quoted(i, 1))
     }
 
     pub fn regex(i: InternedString) -> TerminalString {
-        TerminalString::Literal(TerminalLiteral::Regex(i))
+        TerminalString::Literal(TerminalLiteral::Regex(i, 0))
+    }
+
+    pub fn with_match_precedence(self, p: usize) -> TerminalString {
+        let base_precedence = p * 2;  // Multiply times two since we still want to distinguish between quoted and regex precedence
+        match self {
+            TerminalString::Literal(TerminalLiteral::Quoted(i, _)) =>
+                TerminalString::Literal(TerminalLiteral::Quoted(i, base_precedence+1)),
+            TerminalString::Literal(TerminalLiteral::Regex(i, _)) =>
+                TerminalString::Literal(TerminalLiteral::Regex(i, base_precedence+0)),
+            _ => panic!("cannot set match precedence for {:?}", self)
```

Comment:
Something doesn't feel right to me here. I noticed in a few places that we were converting to/from `TerminalString` and `TerminalLiteral` more than seems necessary. I suspect there is some data structure that is defined in terms of `TerminalString` that actually should be defined in terms of `TerminalLiteral`.